### PR TITLE
Routing command from javascript to other kernels

### PR DESCRIPTION
**Pull request by colombod**

### CODE REVIEW

File: src/Microsoft.DotNet.Interactive/Commands/KernelCommand.cs [hunk 1]
```diff
@@ -5,60 +5,81 @@
...
-        [JsonIgnore] public RoutingSlip RoutingSlip { get; }
+    [JsonIgnore]
+    public IReadOnlyCollection<Uri> RoutingSlip
+    {
+        get
+        {
+            var unique = new HashSet<Uri>();
+            var list = new List<Uri>(Parent?.RoutingSlip ?? Array.Empty<Uri>());
+            list.AddRange(_routingSlip.Where(unique.Add));
+            return list;
```
Comment (jonsequitur) [inline src/Microsoft.DotNet.Interactive/Commands/KernelCommand.cs:63]: Properties shouldn't return a new instance on each call. Is there a reason why this state isn't being built up only once?

File: src/Microsoft.DotNet.Interactive/Commands/KernelCommand.cs [hunk 2]
```diff
@@ -5,60 +5,81 @@
...
+    internal bool? ShouldPublishCompletionEvent { get; set; }
 
-        [JsonIgnore] 
-        public ParseResult KernelChooserParseResult { get; internal set; }
+    [JsonIgnore] 
+    public ParseResult KernelChooserParseResult { get; internal set; }
 
-        [JsonIgnore] public RoutingSlip RoutingSlip { get; }
+    [JsonIgnore]
+    public IReadOnlyCollection<Uri> RoutingSlip
```
Comment (jonsequitur) [inline src/Microsoft.DotNet.Interactive/Commands/KernelCommand.cs:56]: This should be a `RoutingSlip`.

File: src/Microsoft.DotNet.Interactive/Kernel.cs [hunk 3]
```diff
@@ -405,18 +405,21 @@ public void RegisterCommandType<TCommand>()
...
                         default:
-                            await Scheduler.RunAsync(
-                                               c,
-                                               InvokePipelineAndCommandHandler,
-                                               c.SchedulingScope.ToString(),
-                                               cancellationToken: cancellationToken)
-                                           .ContinueWith(t =>
-                                           {
-                                               if (t.IsCanceled)
+                            if (!context.IsFailed)
```
Comment (jonsequitur) [inline src/Microsoft.DotNet.Interactive/Kernel.cs:408]: Why is this check specific to the `default` case?

File: src/Microsoft.DotNet.Interactive/Kernel.cs [hunk 4]
```diff
@@ -529,7 +532,24 @@ internal async Task<KernelCommandResult> InvokePipelineAndCommandHandler(KernelC
...
+                            {
+                                return false;
+                            }
+
+                            if (outer == inner.Parent)
+                            {
+                                return true;
+                            }
+
+                            var isInRoutingSlip = outer.RoutingSlip.Zip(inner.RoutingSlip, (o, i) => o.Equals(i)).All(x => x);
```
Comment (jonsequitur) [inline src/Microsoft.DotNet.Interactive/Kernel.cs:548]: This logic looks like a good candidate to move into `RoutingSlip`.

File: src/Microsoft.DotNet.Interactive/KernelInvocationContext.cs [hunk 5]
```diff
@@ -62,14 +62,18 @@ private KernelInvocationContext(KernelCommand command)
...
-        
+
             _disposables.Add(operation);
         }
 
         public KernelCommand Command { get; }
 
         public bool IsComplete { get; private set; }
 
+        private bool _isFailed;
```
Comment (jonsequitur) [inline src/Microsoft.DotNet.Interactive/KernelInvocationContext.cs:73]: Nitpick: Move field declarations to the top of the class.

File: src/Microsoft.DotNet.Interactive/KernelScheduler.cs [hunk 6]
```diff
@@ -115,32 +117,32 @@ private void ScheduledOperationRunLoop(object _)
...
+                    _currentTopLevelOperation = default;
                     _currentlyRunningOperation = null;
                 }
             }
         }
 
         private void Run(ScheduledOperation operation)
         {
-            _currentTopLevelOperation.Value ??= operation;
+            _currentTopLevelOperation = _currentTopLevelOperation is null? operation.Value : _currentTopLevelOperation;
```
Comment (jonsequitur) [inline src/Microsoft.DotNet.Interactive/KernelScheduler.cs:128]: There's a potential race condition here.

```suggestion
            _currentTopLevelOperation = _currentTopLevelOperation ?? operation.Value;
```

File: src/Microsoft.DotNet.Interactive/RoutingSlip.cs [hunk 7]
```diff
@@ -38,9 +38,9 @@ IEnumerator<Uri> IEnumerable<Uri>.GetEnumerator()
 
     IEnumerator IEnumerable.GetEnumerator()
     {
-        return _uris.GetEnumerator();
+        return ((IEnumerable)_uris).GetEnumerator();
```
Comment (jonsequitur) [inline src/Microsoft.DotNet.Interactive/RoutingSlip.cs:47]: This should not be necessary.

File: src/Microsoft.DotNet.Interactive/RoutingSlip.cs [hunk 8]
```diff
@@ -38,9 +38,9 @@ IEnumerator<Uri> IEnumerable<Uri>.GetEnumerator()
...
     }
 
-    public Uri this[int index] => _uris[index];
+    public Uri this[int i] => _uris[i];
```
Comment (jonsequitur) [inline src/Microsoft.DotNet.Interactive/RoutingSlip.cs:50]: I think `index` is preferred.

File: src/Microsoft.DotNet.Interactive/Commands/KernelCommand.cs (hunk 1 above)
Comment (colombod) [reply src/Microsoft.DotNet.Interactive/Commands/KernelCommand.cs]: changed that to be built when parenting is changed

File: src/Microsoft.DotNet.Interactive/Commands/KernelCommand.cs (hunk 2 above)
Comment (colombod) [reply src/Microsoft.DotNet.Interactive/Commands/KernelCommand.cs]: Done

File: src/Microsoft.DotNet.Interactive/Kernel.cs (hunk 4 above)
Comment (colombod) [reply src/Microsoft.DotNet.Interactive/Kernel.cs:548]: Correct! Done

File: src/Microsoft.DotNet.Interactive/KernelInvocationContext.cs (hunk 5 above)
Comment (colombod) [reply src/Microsoft.DotNet.Interactive/KernelInvocationContext.cs:73]: Done

File: src/Microsoft.DotNet.Interactive/RoutingSlip.cs (hunk 8 above)
Comment (colombod) [reply src/Microsoft.DotNet.Interactive/RoutingSlip.cs:59]: done

File: src/Microsoft.DotNet.Interactive/RoutingSlip.cs (hunk 7 above)
Comment (colombod) [reply src/Microsoft.DotNet.Interactive/RoutingSlip.cs:56]: done

**Review by brettfo — APPROVED** [review]

**Pull request merged**